File: services/iceberg-catalog-service/src/handlers/auth/oauth.rs
//! `POST /iceberg/v1/oauth/tokens` — Iceberg-flavoured OAuth2 token
//! endpoint per the REST Catalog spec § Authentication.
//!
//! Two grant types are supported:
//!
//!   * `client_credentials` — body validated against
//!     `oauth-integration-service`'s `POST /v1/oauth-clients/validate`,
//!     access token issued as a JWT with `iss=foundry-iceberg`,
//!     `aud=iceberg-catalog`, scope claims in `scp` and configurable
//!     TTL (default 1 h).
//!   * `refresh_token` — refresh tokens are short-lived JWTs signed
//!     with the same secret; we accept any valid one and re-issue an
//!     access token with the same scope.

use axum::extract::State;
use axum::http::HeaderMap;
use axum::{Form, Json};
use serde::{Deserialize, Serialize};

use crate::AppState;
use crate::audit;
use crate::handlers::auth::bearer::issue_internal_jwt;
use crate::handlers::errors::ApiError;
use crate::metrics;

#[derive(Debug, Deserialize)]
pub struct OAuthTokenForm {
    pub grant_type: String,
    pub client_id: Option<String>,
    pub client_secret: Option<String>,
    pub scope: Option<String>,
    pub refresh_token: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct OAuthTokenResponse {
    pub access_token: String,
    pub token_type: String,
    pub expires_in: i64,
    pub issued_token_type: String,
    pub scope: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub refresh_token: Option<String>,
}

/// Per spec the catalog accepts both form-encoded bodies and JSON.
/// Foundry's PyIceberg builds a form-encoded payload, so we keep that
/// path on the hot road.
pub async fn issue_token(
    State(state): State<AppState>,
    headers: HeaderMap,
    Form(form): Form<OAuthTokenForm>,
) -> Result<Json<OAuthTokenResponse>, ApiError> {
    match form.grant_type.as_str() {
        "client_credentials" => client_credentials_grant(&state, &headers, form).await,
        "refresh_token" => refresh_token_grant(&state, form).await,
        other => Err(ApiError::BadRequest(format!(
            "unsupported grant_type `{other}`"
        ))),
    }
}

async fn client_credentials_grant(
    state: &AppState,
    headers: &HeaderMap,
    form: OAuthTokenForm,
) -> Result<Json<OAuthTokenResponse>, ApiError> {
    let (client_id, client_secret) = resolve_client_credentials(headers, &form)?;

    // Validate credentials against oauth-integration-service. Treat
    // network failures as 503 so PyIceberg retries with backoff.
    let validate_url = format!(
        "{}/v1/oauth-clients/validate",
        state.iceberg.oauth_integration_url.trim_end_matches('/')
    );
    let validate_body = serde_json::json!({
        "client_id": client_id,
        "client_secret": client_secret,
        "scope": form.scope,
    });

    let response = state
        .iceberg
        .http
        .post(&validate_url)
        .json(&validate_body)
        .send()
        .await
        .map_err(|err| ApiError::Internal(format!("oauth validation request failed: {err}")))?;

    if !response.status().is_success() {
        return Err(ApiError::Forbidden(format!(
            "oauth client credentials rejected (status {})",
            response.status()
        )));
    }

    let scopes = form
        .scope
        .clone()
        .unwrap_or_else(|| "api:iceberg-read api:iceberg-write".to_string());
    let scope_list: Vec<String> = scopes.split_whitespace().map(str::to_string).collect();
    let access_token = issue_internal_jwt(
        &state.iceberg.jwt_config,
        &client_id,
        &state.iceberg.jwt_issuer,
        &state.iceberg.jwt_audience,
        &scope_list,
        state.iceberg.default_token_ttl_secs,
    )?;
    let refresh_token = issue_internal_jwt(
        &state.iceberg.jwt_config,
        &client_id,
        &state.iceberg.jwt_issuer,
        &state.iceberg.jwt_audience,
        &scope_list,
        state.iceberg.default_token_ttl_secs * 24,
    )?;

    metrics::OAUTH_TOKENS_ISSUED
        .with_label_values(&["client_credentials"])
        .inc();
    audit::oauth_token_issued(None, "client_credentials", &scopes);

    Ok(Json(OAuthTokenResponse {
        access_token,
        token_type: "bearer".to_string(),
        expires_in: state.iceberg.default_token_ttl_secs,
        issued_token_type: "urn:ietf:params:oauth:token-type:access_token".to_string(),
        scope: scopes,
        refresh_token: Some(refresh_token),
    }))
}

async fn refresh_token_grant(
    state: &AppState,
    form: OAuthTokenForm,
) -> Result<Json<OAuthTokenResponse>, ApiError> {
    let refresh = form
        .refresh_token
        .ok_or_else(|| ApiError::BadRequest("refresh_token is required".to_string()))?;

    // Inline decode using the iceberg-only path. Reusing the bearer
    // module's helper would couple the surfaces; we instead duplicate
    // the small claim shape so this handler stands alone.
    let claims = decode_refresh(&refresh, &state.iceberg.jwt_config)?;

    let scope_list: Vec<String> = claims.scp.split_whitespace().map(str::to_string).collect();
    let access_token = issue_internal_jwt(
        &state.iceberg.jwt_config,
        &claims.sub,
        &state.iceberg.jwt_issuer,
        &state.iceberg.jwt_audience,
        &scope_list,
        state.iceberg.default_token_ttl_secs,
    )?;

    metrics::OAUTH_TOKENS_ISSUED
        .with_label_values(&["refresh_token"])
        .inc();
    audit::oauth_token_issued(None, "refresh_token", &claims.scp);

    Ok(Json(OAuthTokenResponse {
        access_token,
        token_type: "bearer".to_string(),
        expires_in: state.iceberg.default_token_ttl_secs,
        issued_token_type: "urn:ietf:params:oauth:token-type:access_token".to_string(),
        scope: claims.scp,
        refresh_token: None,
    }))
}

#[derive(Debug, Deserialize)]
struct RefreshClaims {
    sub: String,
    #[serde(default)]
    scp: String,
    #[allow(dead_code)]
    aud: String,
    #[allow(dead_code)]
    exp: i64,
}

fn decode_refresh(
    token: &str,
    _config: &auth_middleware::jwt::JwtConfig,
) -> Result<RefreshClaims, ApiError> {
    use jsonwebtoken::{Algorithm, DecodingKey, Validation};
    let mut validation = Validation::new(Algorithm::HS256);
    validation.validate_exp = true;
    let secret = std::env::var("OPENFOUNDRY_JWT_SECRET")
        .or_else(|_| std::env::var("JWT_SECRET"))
        .unwrap_or_else(|_| "iceberg-catalog-dev-secret".to_string());
    let key = DecodingKey::from_secret(secret.as_bytes());
    // Refresh tokens accept the same audience as access tokens.
    validation.validate_aud = false;
    let data = jsonwebtoken::decode::<RefreshClaims>(token, &key, &validation)
        .map_err(|err| ApiError::Unauthenticated.message(err.to_string()))?;
    Ok(data.claims)
}

trait WithMessage {
    fn message(self, msg: String) -> Self;
}

impl WithMessage for ApiError {
    fn message(self, msg: String) -> Self {
        tracing::debug!(error.detail = %msg, "refresh token rejected");
        self
    }
}

fn resolve_client_credentials(
    headers: &HeaderMap,
    form: &OAuthTokenForm,
) -> Result<(String, String), ApiError> {
    if let (Some(id), Some(secret)) = (form.client_id.clone(), form.client_secret.clone()) {
        return Ok((id, secret));
    }

    // Fallback to RFC 6749 § 2.3.1 HTTP Basic auth header.
    if let Some(value) = headers.get(axum::http::header::AUTHORIZATION) {
        if let Ok(text) = value.to_str() {
            if let Some(b64) = text.strip_prefix("Basic ") {
                use base64::Engine as _;
                if let Ok(decoded) = base64::engine::general_purpose::STANDARD.decode(b64) {
                    if let Ok(decoded) = String::from_utf8(decoded) {
                        if let Some((id, secret)) = decoded.split_once(':') {
                            return Ok((id.to_string(), secret.to_string()));
                        }
                    }
                }
            }
        }
    }

    Err(ApiError::BadRequest(
        "client_id and client_secret required (form or HTTP Basic)".to_string(),
    ))
}

File: services/iceberg-catalog-service/internal/handlers/auth/oauth.go
package auth

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/golang-jwt/jwt/v5"

	"github.com/openfoundry/openfoundry-go/services/iceberg-catalog-service/internal/audit"
)

// OAuthClientValidator is the contract the `client_credentials` branch
// uses to verify caller-supplied credentials. The production
// implementation hits `oauth-integration-service`; tests substitute
// fakes that always allow / always deny.
type OAuthClientValidator interface {
	ValidateClientCredentials(ctx context.Context, clientID, clientSecret, scope string) error
}

// HTTPClientValidator delegates credential checks to
// `oauth-integration-service`'s `POST /v1/oauth-clients/validate`.
// Non-2xx responses are mapped to ErrForbidden so the surfaced 4xx is
// stable for clients.
type HTTPClientValidator struct {
	BaseURL string
	HTTP    *http.Client
}

// ValidateClientCredentials posts the credentials to the integration
// service. Network errors surface as plain errors so the HTTP layer
// returns 5xx (matching the Rust 503 retry contract).
func (v *HTTPClientValidator) ValidateClientCredentials(ctx context.Context, clientID, clientSecret, scope string) error {
	if v == nil {
		return errors.New("oauth validator not configured")
	}
	httpClient := v.HTTP
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	url := strings.TrimRight(v.BaseURL, "/") + "/v1/oauth-clients/validate"
	body, err := json.Marshal(map[string]any{
		"client_id":     clientID,
		"client_secret": clientSecret,
		"scope":         scope,
	})
	if err != nil {
		return fmt.Errorf("marshal validate body: %w", err)
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("build validate request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("oauth validation request failed: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode/100 != 2 {
		return &ErrForbidden{Message: fmt.Sprintf("oauth client credentials rejected (status %d)", resp.StatusCode)}
	}
	return nil
}

// OAuthTokenForm is the spec-compliant request body. Both
// `application/x-www-form-urlencoded` and `application/json` payloads
// decode into this shape.
type OAuthTokenForm struct {
	GrantType    string `json:"grant_type"`
	ClientID     string `json:"client_id"`
	ClientSecret string `json:"client_secret"`
	Scope        string `json:"scope"`
	RefreshToken string `json:"refresh_token"`
}

// OAuthTokenResponse mirrors the spec's TokenResponse JSON shape.
type OAuthTokenResponse struct {
	AccessToken     string `json:"access_token"`
	TokenType       string `json:"token_type"`
	ExpiresIn       int64  `json:"expires_in"`
	IssuedTokenType string `json:"issued_token_type"`
	Scope           string `json:"scope"`
	RefreshToken    string `json:"refresh_token,omitempty"`
}

// IssueTokenHandler returns the HTTP handler for
// `POST /iceberg/v1/oauth/tokens`. Branches on `grant_type` per the
// REST Catalog § Authentication spec.
func IssueTokenHandler(cfg *Config, validator OAuthClientValidator) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		form, err := decodeOAuthForm(r)
		if err != nil {
			writeJSONErr(w, http.StatusBadRequest, err.Error())
			return
		}
		switch form.GrantType {
		case "client_credentials":
			handleClientCredentials(w, r, cfg, validator, form)
		case "refresh_token":
			handleRefreshToken(w, r, cfg, form)
		default:
			writeJSONErr(w, http.StatusBadRequest, fmt.Sprintf("unsupported grant_type `%s`", form.GrantType))
		}
	}
}

func handleClientCredentials(w http.ResponseWriter, r *http.Request, cfg *Config, validator OAuthClientValidator, form OAuthTokenForm) {
	clientID, clientSecret, ok := resolveClientCredentials(r.Header.Get("Authorization"), form)
	if !ok {
		writeJSONErr(w, http.StatusBadRequest, "client_id and client_secret required (form or HTTP Basic)")
		return
	}
	if validator != nil {
		if err := validator.ValidateClientCredentials(r.Context(), clientID, clientSecret, form.Scope); err != nil {
			var forbid *ErrForbidden
			if errors.As(err, &forbid) {
				writeJSONErr(w, http.StatusForbidden, err.Error())
				return
			}
			writeJSONErr(w, http.StatusBadGateway, err.Error())
			return
		}
	}
	scope := form.Scope
	if scope == "" {
		scope = "api:iceberg-read api:iceberg-write"
	}
	scopeList := strings.Fields(scope)
	access, err := IssueInternalJWT(cfg, clientID, cfg.JWTIssuer, cfg.JWTAudience, scopeList, cfg.DefaultTokenTTLSecs)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, err.Error())
		return
	}
	refresh, err := IssueInternalJWT(cfg, clientID, cfg.JWTIssuer, cfg.JWTAudience, scopeList, cfg.DefaultTokenTTLSecs*24)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, err.Error())
		return
	}
	audit.OAuthTokenIssued(nil, "client_credentials", scope)
	writeJSON(w, http.StatusOK, OAuthTokenResponse{
		AccessToken:     access,
		TokenType:       "bearer",
		ExpiresIn:       cfg.DefaultTokenTTLSecs,
		IssuedTokenType: "urn:ietf:params:oauth:token-type:access_token",
		Scope:           scope,
		RefreshToken:    refresh,
	})
}

func handleRefreshToken(w http.ResponseWriter, r *http.Request, cfg *Config, form OAuthTokenForm) {
	if form.RefreshToken == "" {
		writeJSONErr(w, http.StatusBadRequest, "refresh_token is required")
		return
	}
	claims, err := decodeRefreshToken(form.RefreshToken, cfg)
	if err != nil {
		writeJSONErr(w, http.StatusUnauthorized, err.Error())
		return
	}
	scopeList := strings.Fields(claims.Scp)
	access, err := IssueInternalJWT(cfg, claims.Sub, cfg.JWTIssuer, cfg.JWTAudience, scopeList, cfg.DefaultTokenTTLSecs)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, err.Error())
		return
	}
	audit.OAuthTokenIssued(nil, "refresh_token", claims.Scp)
	writeJSON(w, http.StatusOK, OAuthTokenResponse{
		AccessToken:     access,
		TokenType:       "bearer",
		ExpiresIn:       cfg.DefaultTokenTTLSecs,
		IssuedTokenType: "urn:ietf:params:oauth:token-type:access_token",
		Scope:           claims.Scp,
	})
	_ = r // ctx-style placeholder: handler doesn't propagate ctx into JWT decode.
}

func decodeOAuthForm(r *http.Request) (OAuthTokenForm, error) {
	contentType := strings.ToLower(strings.TrimSpace(strings.SplitN(r.Header.Get("Content-Type"), ";", 2)[0]))
	if contentType == "application/json" {
		var f OAuthTokenForm
		body, err := io.ReadAll(r.Body)
		if err != nil {
			return f, fmt.Errorf("read body: %w", err)
		}
		if len(bytes.TrimSpace(body)) == 0 {
			return f, errors.New("empty body")
		}
		if err := json.Unmarshal(body, &f); err != nil {
			return f, fmt.Errorf("invalid json body: %w", err)
		}
		return f, nil
	}
	if err := r.ParseForm(); err != nil {
		return OAuthTokenForm{}, fmt.Errorf("invalid form body: %w", err)
	}
	return OAuthTokenForm{
		GrantType:    r.PostForm.Get("grant_type"),
		ClientID:     r.PostForm.Get("client_id"),
		ClientSecret: r.PostForm.Get("client_secret"),
		Scope:        r.PostForm.Get("scope"),
		RefreshToken: r.PostForm.Get("refresh_token"),
	}, nil
}

func resolveClientCredentials(authHeader string, form OAuthTokenForm) (string, string, bool) {
	if form.ClientID != "" && form.ClientSecret != "" {
		return form.ClientID, form.ClientSecret, true
	}
	if v, ok := strings.CutPrefix(authHeader, "Basic "); ok {
		decoded, err := base64.StdEncoding.DecodeString(strings.TrimSpace(v))
		if err == nil {
			parts := strings.SplitN(string(decoded), ":", 2)
			if len(parts) == 2 {
				return parts[0], parts[1], true
			}
		}
	}
	return "", "", false
}

// decodeRefreshToken decodes the refresh JWT. We don't validate audience
// here: refresh tokens are minted by us, and PyIceberg sometimes calls
// the endpoint cross-environment (which would otherwise mismatch).
func decodeRefreshToken(token string, cfg *Config) (*IcebergClaims, error) {
	parser := jwt.NewParser(jwt.WithValidMethods([]string{"HS256"}))
	claims := &IcebergClaims{}
	parsed, err := parser.ParseWithClaims(token, claims, func(t *jwt.Token) (any, error) {
		return cfg.Secret, nil
	})
	if err != nil {
		return nil, fmt.Errorf("refresh token rejected: %w", err)
	}
	if !parsed.Valid {
		return nil, errors.New("refresh token not valid")
	}
	return claims, nil
}
